File: examples/ios/Examples/Notifications.m
// :replace-start: {
//   "terms": {
//     "NotificationsObjcExamples_": ""
//   }
// }
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>

// :code-block-start: register-an-object-change-listener
@interface NotificationsObjcExamples_Dog : RLMObject
@property NSString *name;
@property int age;
@end

@implementation NotificationsObjcExamples_Dog
@end

RLMNotificationToken *objectNotificationToken = nil;

void objectNotificationExample() {
    NotificationsObjcExamples_Dog *dog = [[NotificationsObjcExamples_Dog alloc] init];
    dog.name = @"Max";
    dog.age = 3;
    
    // Open the default realm
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm transactionWithBlock:^{
        [realm addObject:dog];
    }];

    // Observe object notifications. Keep a strong reference to the notification token
    // or the observation will stop. Invalidate the token when done observing.
    objectNotificationToken = [dog addNotificationBlock:^(BOOL deleted, NSArray<RLMPropertyChange *> * _Nullable changes, NSError * _Nullable error) {
        if (error != nil) {
            NSLog(@"An error occurred: %@", [error localizedDescription]);
            return;
        }
        if (deleted) {
            NSLog(@"The object was deleted.");
            return;
        }
        NSLog(@"Property %@ changed to '%@'",
              changes[0].name,
              changes[0].value);
    }];
    
    // Now update to trigger the notification
    [realm transactionWithBlock:^{
        dog.name = @"Wolfie";
    }];

}
// :code-block-end:

// :code-block-start: register-a-collection-change-listener
@interface CollectionNotificationExampleViewController : UITableViewController
@end

@implementation CollectionNotificationExampleViewController {
    RLMNotificationToken *notificationToken;
}
- (instancetype)init {
    if (self = [super init]) {
        self->notificationToken = nil;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    // Observe RLMResults Notifications
    __weak typeof(self) weakSelf = self;
    self->notificationToken = [[NotificationsObjcExamples_Dog objectsWhere:@"age > 5"]
      addNotificationBlock:^(RLMResults<NotificationsObjcExamples_Dog *> *results, RLMCollectionChange *changes, NSError *error) {
        
        if (error) {
            NSLog(@"Failed to open realm on background worker: %@", error);
            return;
        }

        UITableView *tableView = weakSelf.tableView;
        // Initial run of the query will pass nil for the change information
        if (!changes) {
            [tableView reloadData];
            return;
        }

        // Query results have changed, so apply them to the UITableView
        [tableView performBatchUpdates:^{
            // Always apply updates in the following order: deletions, insertions, then modifications.
            // Handling insertions before deletions may result in unexpected behavior.
            [tableView deleteRowsAtIndexPaths:[changes deletionsInSection:0]
                             withRowAnimation:UITableViewRowAnimationAutomatic];
            [tableView insertRowsAtIndexPaths:[changes insertionsInSection:0]
                             withRowAnimation:UITableViewRowAnimationAutomatic];
            [tableView reloadRowsAtIndexPaths:[changes modificationsInSection:0]
                             withRowAnimation:UITableViewRowAnimationAutomatic];
        } completion:^(BOOL finished) {
            // ...
        }];
    }];
}

- (void)dealloc {
    [self->notificationToken invalidate];
}
@end
// :code-block-end:

@interface NotificationsObjc : XCTestCase
@end

@implementation NotificationsObjc
- (void)tearDown {
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm transactionWithBlock:^{
        [realm deleteAllObjects];
    }];
}

- (void)testObjectNotificationExample {
    objectNotificationExample();
}

- (void)testSilentWrite {
    // :code-block-start: silent-write
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    // Observe realm notifications
    RLMNotificationToken *token = [realm addNotificationBlock:^(RLMNotification  _Nonnull notification, RLMRealm * _Nonnull realm) {
        // ... handle update
    }];
    
    // Later, pass the token in an array to the realm's `-transactionWithoutNotifying:block:` method.
    // Realm will _not_ notify the handler after this write.
    [realm transactionWithoutNotifying:@[token] block:^{
       // ... write to realm
    }];
    
    // Finally
    [token invalidate];
    // :code-block-end:
}

- (void)testRealmNotification {
    // :code-block-start: register-a-realm-change-listener
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    // Observe realm notifications. Keep a strong reference to the notification token
    // or the observation will stop.
    RLMNotificationToken *token = [realm addNotificationBlock:^(RLMNotification  _Nonnull notification, RLMRealm * _Nonnull realm) {
        // `notification` is an enum specifying what kind of notification was emitted.
        // ... update UI ...
    }];
    
    // ...

    // Later, explicitly stop observing.
    [token invalidate];
    // :code-block-end:
}

- (void)testStopWatching {
    // :code-block-start: stop-watching
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    // Observe and obtain token
    RLMNotificationToken *token = [realm addNotificationBlock:^(RLMNotification  _Nonnull notification, RLMRealm * _Nonnull realm) {
        /* ... */
    }];

    // Stop observing
    [token invalidate];
    // :code-block-end:
}

@end

// :replace-end:

File: examples/ios/Examples/Notifications.swift
// :replace-start: {
//   "terms": {
//     "AlternateNotificationExampleKeyPath_": "",
//     "additionalObjectNotificationToken": "objectNotificationToken",
//     "keyPath_": "",
//     "NotificationExample_": "",
//     "NotificationExampleKeyPath_": ""
//   }
// }
import RealmSwift
import XCTest

// Note: the following is not really tested but it makes for a convenient
// full program example.
// :snippet-start: register-an-object-change-listener
// Define the dog class.
class NotificationExample_Dog: Object {
    @Persisted var name = ""
}

var objectNotificationToken: NotificationToken?

func objectNotificationExample() {
    let dog = NotificationExample_Dog()
    dog.name = "Max"

    // Open the default realm.
    let realm = try! Realm()
    try! realm.write {
        realm.add(dog)
    }
    // Observe object notifications. Keep a strong reference to the notification token
    // or the observation will stop. Invalidate the token when done observing.
    objectNotificationToken = dog.observe { change in
        switch change {
        case .change(let object, let properties):
            for property in properties {
                print("Property '\(property.name)' of object \(object) changed to '\(property.newValue!)'")
            }
        case .error(let error):
            print("An error occurred: \(error)")
        case .deleted:
            print("The object was deleted.")
        }
    }

    // Now update to trigger the notification
    try! realm.write {
        dog.name = "Wolfie"
    }
}
// :snippet-end:

// Note: the following is not really tested but it makes for a convenient
// full program example.
// :snippet-start: register-a-keypath-change-listener
// Define the dog class.
class NotificationExampleKeyPath_Dog: Object {
    @Persisted var name = ""
    @Persisted var favoriteToy = ""
    @Persisted var age: Int?
}

var additionalObjectNotificationToken: NotificationToken?

func keyPath_objectNotificationExample() {
    let dog = NotificationExampleKeyPath_Dog()
    dog.name = "Max"
    dog.favoriteToy = "Ball"
    dog.age = 2

    // Open the default realm.
    let realm = try! Realm()
    try! realm.write {
        realm.add(dog)
    }
    // Observe notifications on some of the object's key paths. Keep a strong
    // reference to the notification token or the observation will stop.
    // Invalidate the token when done observing.
    objectNotificationToken = dog.observe(keyPaths: ["favoriteToy", "age"], { change in // :emphasize:
        switch change {
        case .change(let object, let properties):
            for property in properties {
                print("Property '\(property.name)' of object \(object) changed to '\(property.newValue!)'")
            }
        case .error(let error):
            print("An error occurred: \(error)")
        case .deleted:
            print("The object was deleted.")
        }
    })

    // Now update to trigger the notification
    try! realm.write {
        dog.favoriteToy = "Frisbee"
    }
    // When you specify one or more key paths, changes to other properties
    // do not trigger notifications. In this example, changing the "name"
    // property does not trigger a notification.
    try! realm.write {
        dog.name = "Maxamillion"
    }
}
// :snippet-end:

// :snippet-start: alternate-dog-class-for-keypaths
class AlternateNotificationExampleKeyPath_Dog: Object {
    @Persisted var name = ""
    @Persisted var siblings: List<AlternateNotificationExampleKeyPath_Dog> // :emphasize:
    @Persisted var age: Int?
}
// :snippet-end:

// :snippet-start: register-a-collection-change-listener
class CollectionNotificationExampleViewController: UITableViewController {
    var notificationToken: NotificationToken?

    override func viewDidLoad() {
        super.viewDidLoad()
        let realm = try! Realm()
        let results = realm.objects(NotificationExample_Dog.self)

        // Observe collection notifications. Keep a strong
        // reference to the notification token or the
        // observation will stop.
        notificationToken = results.observe { [weak self] (changes: RealmCollectionChange) in
            guard let tableView = self?.tableView else { return }
            switch changes {
            case .initial:
                // Results are now populated and can be accessed without blocking the UI
                tableView.reloadData()
            case .update(_, let deletions, let insertions, let modifications):
                // Query results have changed, so apply them to the UITableView
                tableView.performBatchUpdates({
                    // Always apply updates in the following order: deletions, insertions, then modifications.
                    // Handling insertions before deletions may result in unexpected behavior.
                    tableView.deleteRows(at: deletions.map({ IndexPath(row: $0, section: 0)}),
                                         with: .automatic)
                    tableView.insertRows(at: insertions.map({ IndexPath(row: $0, section: 0) }),
                                         with: .automatic)
                    tableView.reloadRows(at: modifications.map({ IndexPath(row: $0, section: 0) }),
                                         with: .automatic)
                }, completion: { finished in
                    // ...
                })
            case .error(let error):
                // An error occurred while opening the Realm file on the background worker thread
                fatalError("\(error)")
            }
        }
    }
}
// :snippet-end:

class Notifications: XCTestCase {
    override func tearDown() {
        let realm = try! Realm()
        try! realm.write {
            realm.deleteAll()
        }
    }

    func testSilentWrite() {
        // :snippet-start: silent-write
        let realm = try! Realm()

        // Observe realm notifications
        let token = realm.observe { notification, realm in
            // ... handle update
        }

        // Later, pass the token in an array to the realm.write(withoutNotifying:)
        // method to write without sending a notification to that observer.
        try! realm.write(withoutNotifying: [token]) {
            // ... write to realm
        }

        // Finally
        token.invalidate()
        // :snippet-end:
    }

    func testRealmNotification() {
        class VC {
            func updateUI() {}
        }
        let viewController = VC()
        // :snippet-start: register-a-realm-change-listener
        let realm = try! Realm()

        // Observe realm notifications. Keep a strong reference to the notification token
        // or the observation will stop.
        let token = realm.observe { notification, realm in
            // `notification` is an enum specifying what kind of notification was emitted
            viewController.updateUI()
        }

        // ...

        // Later, explicitly stop observing.
        token.invalidate()
        // :snippet-end:
    }

    func testStopWatching() {
        // :snippet-start: stop-watching
        let realm = try! Realm()

        // Observe and obtain token
        let token = realm.observe { notification, realm in /* ... */ }

        // Stop observing
        token.invalidate()
        // :snippet-end:
    }
}
// :replace-end:
